examples/dmd/wave_equation.cpp

```diff
@@ -0,0 +1,621 @@
+         cout << "GLVis visualization disabled.\n";
+      }
+      else
+      {
+         sout.precision(precision);
+         sout << "solution\n" << *mesh << dudt_gf;
+         sout << flush;
+      }
+   }
+
+   StopWatch fom_timer, dmd_training_timer, dmd_prediction_timer;
+   fom_timer.Start();
+
+   // 8. Perform time-integration (looping over the time iterations, ti, with a
+   //    time-step dt).
+   ode_solver->Init(oper);
+   double t = 0.0;
+
+   vector<double> ts;
+
+   fom_timer.Stop();
+
+   dmd_training_timer.Start();
+   
+   int curr_window = 0;
+   vector<CAROM::DMD*> dmd_u;
+   dmd_u.push_back(new CAROM::DMD(u_gf.GetTrueVector().Size(), dt));
+   dmd_u[curr_window]->takeSample(u_gf.GetTrueVector(), t);
+   ts.push_back(t);
+
+
+   dmd_training_timer.Stop();
+
+   u_gf.SetFromTrueDofs(u);
+
+
+   bool last_step = false;
+   for (int ti = 1; !last_step; ti++)
+   {
+      fom_timer.Start();
+      if (t + dt >= t_final - dt/2)
+      {
+         last_step = true;
+      }
+
+
+      ode_solver->Step(u, dudt, t, dt);
+      fom_timer.Stop();
+
+      dmd_training_timer.Start();
+      u_gf.SetFromTrueDofs(u);
+      dmd_u[curr_window]->takeSample(u_gf.GetTrueVector(), t);
+
+      if (last_step || (ti % windowNumSamples) == 0)
+      {
+         cout << "step " << ti << ", t = " << t << endl;
+
+         if(rdim != -1){
+            
+            if(myid == 0){
+
+               std::cout << "Creating DMD with rdim: " << rdim << std::endl;
+            }
+            
+            dmd_u[curr_window]->train(rdim);
+         }
+         
+         else if(ef != -1){
+            if(myid == 0){
+               std::cout << "Creating DMD with energy fraction: " << ef << std::endl;
+            }
+            dmd_u[curr_window]->train(ef);
+         }
+
+         if(!last_step){
+             curr_window++;
+             dmd_u.push_back(new CAROM::DMD(u_gf.GetTrueVector().Size(), dt));
+             dmd_u[curr_window]->takeSample(u_gf.GetTrueVector(), t);
+         }
+
+      }
+      ts.push_back(t);
+      dmd_training_timer.Stop();
+       
+     if (last_step || (ti % vis_steps) == 0)
+        {
+            
+            u_gf.SetFromTrueDofs(u);
+            dudt_gf.SetFromTrueDofs(dudt);
+
+            if (visualization)
+            {
+               sout << "solution\n" << *mesh << u_gf << flush;
+            }
+
+            if (visit)
+            {
+               visit_dc.SetCycle(ti);
+               visit_dc.SetTime(t);
+               visit_dc.Save();
+            }
+        }
+       oper.SetParameters(u);
+
+   }
+
+   // 9. Save the final solution. This output can be viewed later using GLVis:
+   //    "glvis -m wave_equation.mesh -g wave_equation-final.gf".
+   {
+      ofstream osol("wave_equation-final.gf");
+      osol.precision(precision);
+      u_gf.Save(osol);
+      dudt_gf.Save(osol);
+   }
+
+   // 10. Calculate the DMD modes.
+    if (myid == 0 && rdim != -1 && ef != -1)
```

Comment:
It looks like `ef != -1` is a copy and paste mistake.

Reply:
@dylan-copeland I rewrote this section of code in the process of implementing changes to this PR. Either `rdim` or `ef` must be set.

Reply:
My point is that `-1` is not a reasonable value to check. I think this was an existing mistake in libROM that was copied and pasted. `ef` is a positive `double`.

Reply:
The convention across libROM seems to denote `-1` as the absence of a value for `ef`, as opposed to a legitimate value.

Reply:
It seems to be a mistake in libROM rather than a convention. The default value is `0.9999`, so `-1` does not mean the parameter was not set. We could make the default value `-1.0` and use the condition `ef > 0.0` to check that a valid value was defined by the user. This would require the user to set `-ef` when desired. We usually have the default value `0.9999` so that the user does not have to specify `-ef`. That seems to be the most convenient option, which means there should be no comparison of `ef` to `-1`.